ex/loqui/lib/loqui/client.ex

```diff
@@ -0,0 +1,358 @@
+    :ok = :gen_tcp.close(sock)
+    case info do
+      {:close, from} ->
+        Connection.reply(from, :ok)
+        {:stop, :normal, %{state | sock: nil}}
+
+      {:error, reason}  ->
+        {:stop, reason, %{state | sock: nil}}
+    end
+  end
+
+  # Connection callbacks
+  def handle_call(:ping, caller, %{sock: sock}=state) do
+    {next_seq, state} = next_sequence(state)
+    :gen_tcp.send(sock, Frames.ping(0, next_seq))
+
+    {:noreply, %State{state | waiters: Map.put(state.waiters, next_seq, caller)}}
+  end
+
+  def handle_call(_, _from, %State{sequence: :go_away}=state) do
+    {:reply, {:error, :remote_went_away}, state}
+  end
+
+  def handle_call(:close, from, %State{sock: sock}=s) do
+    go_away_packet = Frames.goaway(0, 0, "Closing")
+    :gen_tcp.send(sock, go_away_packet)
+
+    {:disconnect, {:close, from}, s}
+  end
+
+  def handle_call({:request, payload}, caller, %State{sock: sock, codec: codec}=state) do
+    {next_seq, state} = next_sequence(state)
+    encoded_payload = codec.encode(payload)
+    :gen_tcp.send(sock, Frames.request(0, next_seq, encoded_payload))
+
+    {:noreply, %State{state | waiters: Map.put(state.waiters, next_seq, caller)}}
+  end
+
+  def handle_cast(_, %State{sequence: :go_away}=state),
+    do: {:noreply, state}
+
+  def handle_cast({:push, payload}, %State{sock: sock, codec: codec}=state) do
+    encoded_payload = codec.encode(payload)
+    :gen_tcp.send(sock, Frames.push(0, encoded_payload))
+
+    {:noreply, state}
+  end
+
+  def handle_info({:tcp, socket, data}, %State{sock: socket}=state) do
+    make_active_once(state)
+    with {:ok, parsed_packets, leftover_data} <- Parser.parse(state.buffer, data) do
+      state = handle_packets(parsed_packets, state)
+
+      {:noreply, %State{state | buffer: leftover_data}}
+    else
+
+      {:error, {:need_more_data, data}} ->
+        {:noreply, %State{buffer: data }}
+    end
+  end
+
+  def handle_info({:tcp_closed, _socket}, _state) do
+    {:stop, :tcp_closed, nil}
+  end
+
+  def handle_info({:tcp_error, _, _}, _state) do
+    {:stop, :tcp_closed, nil}
+  end
+
+  def handle_info({:close_go_away, go_away_code, go_away_data}, %State{sequence: :go_away, waiters: waiters}=state) do
+    # tell the waiters that the remote end went away, then close.
+    err = {:error, {:remote_went_away, go_away_code, go_away_data}}
+    Enum.each(waiters, fn {_, waiter} ->
+      Connection.reply(waiter, err)
+    end)
+
+    {:disconnect, err, state}
+  end
+
+  def handle_info(:send_ping, %State{sock: sock, last_ping: nil}=state) do
+    {next_seq, state} = next_sequence(state)
+
+    :gen_tcp.send(sock, Frames.ping(0, next_seq))
+    new_state = state
+      |> schedule_ping
+      |> Map.put(:last_ping, {next_seq, :erlang.system_time(:milli_seconds)})
+
+    {:noreply, new_state}
+  end
+  def handle_info(:send_ping, %State{last_ping: {sequence, _ts}}=state) do
+    {:stop, {:error, {:pong_not_received, sequence}}, state}
```

Comment:
Yes, the question is if we should account for that case. It feels like we have an overloaded server there.